source-firestore/pull.go

```diff
@@ -0,0 +1,322 @@
+
+	for {
+		var snapshot, err = snapshotsIterator.Next()
+
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		for _, change := range snapshot.Changes {
+			if change.Kind == firestore.DocumentAdded || change.Kind == firestore.DocumentModified {
+
+				// Make sure this document matches the binding path
+				var index = findBindingForCollectionPath(change.Doc.Ref.Parent.Path, c.Bindings)
+				log.WithFields(log.Fields{
+					"kind":    change.Kind,
+					"path":    change.Doc.Ref.Path,
+					"p":       collectionToResourcePath(change.Doc.Ref.Parent.Path),
+					"binding": index,
+				}).Debug("received change")
+				if index == -1 {
+					continue
+				}
+
+				var doc = change.Doc
+				var data = doc.Data()
+				data[firestore.DocumentID] = doc.Ref.ID
+				data[PATH_FIELD] = doc.Ref.Path
+				docJson, err := json.Marshal(data)
+				if err != nil {
+					return err
+				}
+				log.WithFields(log.Fields{
+					"doc": string(docJson),
+					"id":  change.Doc.Ref.ID,
+				}).Debug("received change doc")
+				c.Output <- emitRequest{
+					requestType: EncodeDocument,
+					binding:     uint32(index),
+					arena:       docJson,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *capture) FullScan(ctx context.Context, lastScan time.Time, scanInterval time.Duration, query firestore.Query) error {
+	var nextScan = time.Until(lastScan.Add(scanInterval))
+	// Wait until next scan
+	time.Sleep(nextScan)
+
+	var docsIterator = query.Documents(ctx)
+
+	var scanTime = time.Now()
+	for {
+		var doc, err = docsIterator.Next()
+
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		var index = findBindingForCollectionPath(doc.Ref.Parent.Path, c.Bindings)
+		if index == -1 {
+			continue
+		}
+
+		var data = doc.Data()
+		data[firestore.DocumentID] = doc.Ref.ID
+		data[PATH_FIELD] = doc.Ref.Path
+		docJson, err := json.Marshal(data)
+		if err != nil {
+			return err
+		}
+		c.Output <- emitRequest{
+			requestType: EncodeDocument,
+			binding:     uint32(index),
+			arena:       docJson,
+		}
+	}
+
+	// If we are not tailing, we close the channel after the full scan
+	// to allow the encoding routine to finish so the connector can exit
+	if !c.Tail {
+		return nil
+	}
+
+	return c.FullScan(ctx, scanTime, scanInterval, query)
+}
+
+func (c *capture) EmitDocuments() error {
+	for {
+		emitDoc := <-c.Output
+
+		if emitDoc.requestType == EncodeDocument {
+			var msg = &pc.PullResponse{
+				Documents: &pc.Documents{
+					Binding:  emitDoc.binding,
+					Arena:    emitDoc.arena,
+					DocsJson: []pf.Slice{{Begin: 0, End: uint32(len(emitDoc.arena))}},
+				},
+			}
+			if err := c.Stream.Send(msg); err != nil {
+				return err
+			}
+			if err := c.Stream.Send(&pc.PullResponse{
+				Checkpoint: &pf.DriverCheckpoint{
+					DriverCheckpointJson: []byte("{}"),
+					Rfc7396MergePatch:    true,
+				},
+			}); err != nil {
+				return err
+			}
+		} else if emitDoc.requestType == StateUpdate {
```

Comment:
I'm not seeing where this is produced into the channel?

Reply:
You are right, it isn't anymore, I can remove this part 👍🏽